File: leetcodepractice/sort-an-array/sort-an-array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortArray = function(nums, func) {
    if (nums.length <= 1) return nums;

    if (!func){
        func = (left, right) => {
            return left < right ? -1 : left > right ? 1 : 0;
        }
    }

    const midpoint = Math.floor(nums.length / 2);
    const left = sortArray(nums.slice(0, midpoint), func);
    const right = sortArray(nums.slice(midpoint), func);
    return merger(left, right, func);
};

const merger = (left, right, func) => {
    let merged = []
    
    while (left.length && right.length){
        switch(func(left[0], right[0])){
            case -1:
                merged.push(left.shift());
                break;
            case 0:
                merged.push(left.shift());
                break;
            case 1:
                merged.push(right.shift());
                break;
        }
    }

    merged = merged.concat(left, right)
    return merged;
}

File: leetcodepractice/948-sort-an-array/sort-an-array.rb
# @param {Integer[]} nums
# @return {Integer[]}
def sort_array(nums)
    return nums if nums.length <= 1

    mid = nums.length / 2

    left = nums[0...mid]
    right = nums[mid...nums.length]

    sorted_left = sort_array(left)
    sorted_right = sort_array(right)

    return merged(sorted_left, sorted_right)
end

def merged(left, right)
    merged = []

    while left.length > 0 && right.length > 0 do
        case
        when left[0] < right[0]
            merged.push(left.shift)
        when left[0] == right[0]
            merged.push(left.shift)
        else
            merged.push(right.shift)
        end
    end

    return merged.concat(left, right)
end